From d8407c352b9e12bbb0049a0d2771050199918235 Mon Sep 17 00:00:00 2001
From: Brendan King <Brendan.King@imgtec.com>
Date: Wed, 3 Jan 2024 15:40:04 +0000
Subject: [PATCH 62/71] egl/wayland: always get server GPU using wl_drm

The IMG DDK uses the display driver to allocate scanout memory. The
device returned by wl_drm is typically the display driver, whereas
dmabuf feedback will return a GPU render node.

Upstream-Status: Inappropriate [oe specific]

---
 src/egl/drivers/dri2/egl_dri2.c         | 10 +++
 src/egl/drivers/dri2/egl_dri2.h         |  4 ++
 src/egl/drivers/dri2/platform_wayland.c | 83 +++++++++++++++----------
 3 files changed, 65 insertions(+), 32 deletions(-)

diff --git a/src/egl/drivers/dri2/egl_dri2.c b/src/egl/drivers/dri2/egl_dri2.c
index 9dac8775648..fe36640e164 100644
--- a/src/egl/drivers/dri2/egl_dri2.c
+++ b/src/egl/drivers/dri2/egl_dri2.c
@@ -1429,6 +1429,7 @@ dri2_display_destroy(_EGLDisplay *disp)
 
 #ifdef HAVE_WAYLAND_PLATFORM
    free(dri2_dpy->device_name);
+   free(dri2_dpy->server_device_name);
 #endif
 
 #ifdef HAVE_ANDROID_PLATFORM
@@ -1453,6 +1454,12 @@ dri2_display_destroy(_EGLDisplay *disp)
       break;
    }
 
+#ifdef HAVE_WAYLAND_PLATFORM
+   if (dri2_dpy->fd_server_gpu >= 0 &&
+       dri2_dpy->fd_server_gpu != dri2_dpy->fd_display_gpu &&
+       dri2_dpy->fd_server_gpu != dri2_dpy->fd_render_gpu)
+      close(dri2_dpy->fd_server_gpu);
+#endif
    if (dri2_dpy->fd_display_gpu >= 0 &&
        dri2_dpy->fd_render_gpu != dri2_dpy->fd_display_gpu)
       close(dri2_dpy->fd_display_gpu);
@@ -1483,6 +1490,9 @@ dri2_display_create(void)
 
    dri2_dpy->fd_render_gpu = -1;
    dri2_dpy->fd_display_gpu = -1;
+#ifdef HAVE_WAYLAND_PLATFORM
+   dri2_dpy->fd_server_gpu = -1;
+#endif
 
 #ifdef HAVE_DRI3_MODIFIERS
    dri2_dpy->dri3_major_version = -1;
diff --git a/src/egl/drivers/dri2/egl_dri2.h b/src/egl/drivers/dri2/egl_dri2.h
index 3900f94b334..b63bdd6a70f 100644
--- a/src/egl/drivers/dri2/egl_dri2.h
+++ b/src/egl/drivers/dri2/egl_dri2.h
@@ -336,6 +336,10 @@ struct dri2_egl_display {
    uint32_t capabilities;
    char *device_name;
    bool is_render_node;
+   struct wl_drm *wl_client_drm;
+   int fd_server_gpu;
+   char *server_device_name;
+   bool server_authenticated;
 #endif
 
 #ifdef HAVE_NULL_PLATFORM
diff --git a/src/egl/drivers/dri2/platform_wayland.c b/src/egl/drivers/dri2/platform_wayland.c
index d67924d9978..271981a6fdd 100644
--- a/src/egl/drivers/dri2/platform_wayland.c
+++ b/src/egl/drivers/dri2/platform_wayland.c
@@ -1554,7 +1554,7 @@ dri2_wl_get_display_fd(void *loaderPrivate)
    _EGLDisplay *disp = loaderPrivate;
    struct dri2_egl_display *dri2_dpy = dri2_egl_display(disp);
 
-   return dri2_dpy->fd_display_gpu;
+   return dri2_dpy->fd_server_gpu;
 }
 
 static const __DRIimageLoaderExtension image_loader_extension = {
@@ -1962,31 +1962,31 @@ drm_handle_device(void *data, struct wl_drm *drm, const char *device)
    struct dri2_egl_display *dri2_dpy = data;
    drm_magic_t magic;
 
-   dri2_dpy->device_name = strdup(device);
-   if (!dri2_dpy->device_name)
+   dri2_dpy->server_device_name = strdup(device);
+   if (!dri2_dpy->server_device_name)
       return;
 
-   dri2_dpy->fd_render_gpu = loader_open_device(dri2_dpy->device_name);
-   if (dri2_dpy->fd_render_gpu == -1) {
+   dri2_dpy->fd_server_gpu = loader_open_device(dri2_dpy->server_device_name);
+   if (dri2_dpy->fd_server_gpu == -1) {
       _eglLog(_EGL_WARNING, "wayland-egl: could not open %s (%s)",
-              dri2_dpy->device_name, strerror(errno));
-      free(dri2_dpy->device_name);
-      dri2_dpy->device_name = NULL;
+              dri2_dpy->server_device_name, strerror(errno));
+      free(dri2_dpy->server_device_name);
+      dri2_dpy->server_device_name = NULL;
       return;
    }
 
-   if (drmGetNodeTypeFromFd(dri2_dpy->fd_render_gpu) == DRM_NODE_RENDER) {
-      dri2_dpy->authenticated = true;
+   if (drmGetNodeTypeFromFd(dri2_dpy->fd_server_gpu) == DRM_NODE_RENDER) {
+      dri2_dpy->server_authenticated = true;
    } else {
-      if (drmGetMagic(dri2_dpy->fd_render_gpu, &magic)) {
-         close(dri2_dpy->fd_render_gpu);
-         dri2_dpy->fd_render_gpu = -1;
-         free(dri2_dpy->device_name);
-         dri2_dpy->device_name = NULL;
+      if (drmGetMagic(dri2_dpy->fd_server_gpu, &magic)) {
+         close(dri2_dpy->fd_server_gpu);
+         dri2_dpy->fd_server_gpu = -1;
+         free(dri2_dpy->server_device_name);
+         dri2_dpy->server_device_name = NULL;
          _eglLog(_EGL_WARNING, "wayland-egl: drmGetMagic failed");
          return;
       }
-      wl_drm_authenticate(dri2_dpy->wl_drm, magic);
+      wl_drm_authenticate(dri2_dpy->wl_client_drm, magic);
    }
 }
 
@@ -2015,7 +2015,7 @@ drm_handle_authenticated(void *data, struct wl_drm *drm)
 {
    struct dri2_egl_display *dri2_dpy = data;
 
-   dri2_dpy->authenticated = true;
+   dri2_dpy->server_authenticated = true;
 }
 
 static const struct wl_drm_listener drm_listener = {
@@ -2065,10 +2065,10 @@ static const struct zwp_linux_dmabuf_v1_listener dmabuf_listener = {
 static void
 wl_drm_bind(struct dri2_egl_display *dri2_dpy)
 {
-   dri2_dpy->wl_drm =
+   dri2_dpy->wl_client_drm =
       wl_registry_bind(dri2_dpy->wl_registry, dri2_dpy->wl_drm_name,
                        &wl_drm_interface, dri2_dpy->wl_drm_version);
-   wl_drm_add_listener(dri2_dpy->wl_drm, &drm_listener, dri2_dpy);
+   wl_drm_add_listener(dri2_dpy->wl_client_drm, &drm_listener, dri2_dpy);
 }
 
 static void
@@ -2344,9 +2344,31 @@ dri2_wl_add_configs_for_visuals(_EGLDisplay *disp, bool allow_preserve)
    return (count != 0);
 }
 
+static bool
+dri2_initialize_wayland_wl_drm_extension(struct dri2_egl_display *dri2_dpy)
+{
+   /* wl_drm not advertised by compositor, so can't continue */
+   if (dri2_dpy->wl_drm_name == 0)
+      return false;
+   wl_drm_bind(dri2_dpy);
+
+   if (dri2_dpy->wl_client_drm == NULL)
+      return false;
+   if (roundtrip(dri2_dpy) < 0 || dri2_dpy->fd_server_gpu == -1)
+      return false;
+
+   if (!dri2_dpy->server_authenticated &&
+       (roundtrip(dri2_dpy) < 0 || !dri2_dpy->server_authenticated))
+      return false;
+
+   return true;
+}
+
 static bool
 dri2_initialize_wayland_drm_extensions(struct dri2_egl_display *dri2_dpy)
 {
+   bool have_wl_drm;
+
    /* Get default dma-buf feedback */
    if (dri2_dpy->wl_dmabuf &&
        zwp_linux_dmabuf_v1_get_version(dri2_dpy->wl_dmabuf) >=
@@ -2368,23 +2390,20 @@ dri2_initialize_wayland_drm_extensions(struct dri2_egl_display *dri2_dpy)
       dmabuf_feedback_format_table_fini(&dri2_dpy->format_table);
    }
 
+   have_wl_drm = dri2_initialize_wayland_wl_drm_extension(dri2_dpy);
+
    /* We couldn't retrieve a render node from the dma-buf feedback (or the
     * feedback was not advertised at all), so we must fallback to wl_drm. */
    if (dri2_dpy->fd_render_gpu == -1) {
-      /* wl_drm not advertised by compositor, so can't continue */
-      if (dri2_dpy->wl_drm_name == 0)
+      if (!have_wl_drm)
          return false;
-      wl_drm_bind(dri2_dpy);
 
-      if (dri2_dpy->wl_drm == NULL)
-         return false;
-      if (roundtrip(dri2_dpy) < 0 || dri2_dpy->fd_render_gpu == -1)
-         return false;
-
-      if (!dri2_dpy->authenticated &&
-          (roundtrip(dri2_dpy) < 0 || !dri2_dpy->authenticated))
-         return false;
+      dri2_dpy->wl_drm = dri2_dpy->wl_client_drm;
+      dri2_dpy->fd_render_gpu = dri2_dpy->fd_server_gpu;
+      dri2_dpy->device_name = strdup(dri2_dpy->server_device_name);
+      dri2_dpy->authenticated = dri2_dpy->server_authenticated;
    }
+
    return true;
 }
 
@@ -3105,8 +3124,8 @@ void
 dri2_teardown_wayland(struct dri2_egl_display *dri2_dpy)
 {
    dri2_wl_formats_fini(&dri2_dpy->formats);
-   if (dri2_dpy->wl_drm)
-      wl_drm_destroy(dri2_dpy->wl_drm);
+   if (dri2_dpy->wl_client_drm)
+      wl_drm_destroy(dri2_dpy->wl_client_drm);
    if (dri2_dpy->wl_dmabuf)
       zwp_linux_dmabuf_v1_destroy(dri2_dpy->wl_dmabuf);
    if (dri2_dpy->wl_shm)
-- 
2.34.1

